api/v1beta1/cloudstackmachine_types.go

```diff
@@ -85,6 +88,18 @@ type CloudStackResourceIdentifier struct {
 	Name string `json:"name,omitempty"`
 }
 
+type CloudStackResourceDiskOffering struct {
+	CloudStackResourceIdentifier `json:",inline"`
+	// mount point the data disk uses to mount. The actual partition, mkfs and mount are done by cloud-init generated by kubeadmConfig.
+	MountPath string `json:"mountPath"`
+	// device name of data disk
+	Device string `json:"device"`
+	// filesystem used by data disk
+	Filesystem string `json:"filesystem"`
+	// label of data disk
+	Label string `json:"label"`
```

Comment:
are all of these attributes required? it seems like device name and label might be mutually exclusive, no? Maybe some more detail would be helpful in the descriptions, since I'm not really sure why a disk offering would need a "label" or "device name".

Reply:
I updated the comments for those new fields. Basically, 
device is the device name used when parition a disk, like /dev/vdb which is a device name in linux.
filesystem is the file system used when mkfs (format) a partition, like ext4, xfs
label is a parameter when format disk (mkfs) and label a partition, it is a recommended way to use when mount fs into a directory in linux.

That being said, other than the disk offering ID/Name which is used by CAPC when deploy VM, other parameters are not used by CAPC (device, filesystem, mountPath, label) those parameters are used by kubeadmconfig spec to generate cloud-init script. 

The reason why these parameters are added in cloudstack machine is that eks-a needs to use cloudstack machine spec to re-construct eks-a template to determine if a upgrade is needed. Without these parameters added in CAPC, eks-a will not know if disk offering config changes or not, and could not make a decision to upgrade. 

That being said, if customer uses CAPC directly without eks-a, the deployed VM will have a data disk available when booting, but no partition or file system applied to data disk, and nothing is mounted.

Let CAPC to parition disk, install filesystem and mount to a directory is an alternative, and our implementation uses kubeadmconfig. 

Here is a proposal in 2020 about similar requirements  
https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/proposals/20200423-etcd-data-disk.md